---
title: "Learning the BAM format"
date: "`r Sys.Date()`"
output:
  md_document:
    variant: markdown_github
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

SAMtools provides various (sub)tools for manipulating alignments in the SAM/BAM format. The SAM (Sequence Alignment/Map) format (BAM is just the binary form of SAM) is currently the _de facto_ standard for storing large nucleotide sequence alignments. If you are dealing with high-throughput sequencing data, at some point you will probably have to deal with SAM/BAM files, so familiarise yourself with them! For the latest information, please refer to the [release notes](https://github.com/samtools/samtools/releases).

The examples below use the `ERR188273_chrX.bam` BAM file generated as per https://github.com/davetang/rnaseq using the HISAT2 + StringTie2 RNA-seq pipeline.

# Installing SAMtools

For installing SAMtools, I recommend using the Bioconda [samtools](https://anaconda.org/bioconda/samtools) package. I also recommend using [Miniconda](https://docs.conda.io/en/latest/miniconda.html) instead of Anaconda. I wrote a [short introduction to Conda](https://davetang.github.io/reproducible_bioinformatics/conda.html) if you want to find learn more.

Once you have installed Miniconda, it is easy to install SAMtools.

```bash
conda install -c bioconda samtools
```

# Basic usage

If you run SAMtools on the terminal without any parameters or with `--help`, all the available utilities are listed:

```{bash engine.opts='-l'}
samtools --help
```

# Viewing

Use [bioSyntax](https://github.com/bioSyntax/bioSyntax) to prettify your output.

```bash
samtools view aln.bam | sam-less
```

![bioSyntax](img/sam_less.png)

# Converting a SAM file to a BAM file

A BAM file is just a SAM file but stored in binary format; you should always convert your SAM files into BAM as BAM files are smaller in size and are faster to manipulate.

Since I don't have a SAM file in the example folder, let's first create one and check out the first ten lines. Note: remember to use `-h` to ensure the SAM file contains the sequence header information. Generally, I recommend storing only sorted BAM files as they use less disk space and are faster to process.

```{bash engine.opts='-l'}
samtools view -h eg/ERR188273_chrX.bam > eg/ERR188273_chrX.sam
```

First notice that the SAM file is much larger than the BAM file.

```{bash engine.opts='-l'}
ls -lh eg/ERR188273_chrX.bam eg/ERR188273_chrX.sam
```

We can use `head` to view a SAM file.

```{bash engine.opts='-l'}
head eg/ERR188273_chrX.sam
```

The lines starting with the "@" sign contains the header information. The @SQ tag is the reference sequence dictionary; SN refers to the reference sequence name and LN refers to the reference sequence length. If you don't see lines starting with the "@" sign, the header information is most likely missing. If the @SQ header is absent from the SAM file use the command below, where ref.fa is the reference fasta file used to map the reads, to generate @SQ information.

```bash
samtools view -bT sequence/ref.fa aln.sam > aln.bam
```

If the header information is available, we can convert a SAM file into BAM by using `samtools view -b`. In the newer version of SAMtools the input format is autodetected, so we no longer need the `-S` parameter.

```{bash engine.opts='-l'}
samtools view -b eg/ERR188273_chrX.sam > eg/my.bam
```

# Converting a BAM file to a CRAM file

Use `samtools view` with the `-T` and `-C` arguments to convert a BAM file into CRAM.

```{bash engine.opts='-l'}
samtools view -T /Users/dtang/github/rnaseq/raw/chrX_data/genome/chrX.fa -C -o eg/ERR188273_chrX.cram eg/ERR188273_chrX.bam

ls -lh eg/ERR188273_chrX.[sbcr]*am
```

You can use `samtools view` to view a CRAM file.

```{bash engine.opts='-l'}
samtools view eg/ERR188273_chrX.cram | head
```

I have an [old blog post](https://davetang.org/muse/2014/09/26/bam-to-cram/) on the CRAM format.

# Sorting a SAM/BAM file

Always sort your SAM/BAM files; many downstream programs only take sorted BAM files. In SAMtools version 1.3 or newer, you can sort a SAM file directly.

```{bash engine.opts='-l'}
samtools sort eg/ERR188273_chrX.sam -o eg/sorted.bam
```

You should use use additional threads by specifying `-@ 4` (using 4 threads) to speed up sorting.

```{bash engine.opts='-l'}
time samtools sort eg/ERR188273_chrX.sam -o eg/sorted.bam
time samtools sort -@ 4 eg/ERR188273_chrX.sam -o eg/sorted.bam
```

# Creating a BAM index file

Various tools require BAM index files, such as IGV, which is a program for visualising a BAM file.

```{bash engine.opts='-l'}
samtools index eg/ERR188273_chrX.bam
```

# Filtering unmapped reads

Use `-F 4` to filter out unmapped reads. Use the `flags` subcommand to find out what a flag represents.

```{bash engine.opts='-l'}
samtools flags 4
samtools view -F 4 -b eg/ERR188273_chrX.bam > eg/ERR188273_chrX.mapped.bam
```

Use `-f 4` to keep only unmapped reads.

```{bash engine.opts='-l'}
samtools view -f 4 -b eg/ERR188273_chrX.bam > eg/ERR188273_chrX.unmapped.bam
```

# Extracting entries mapping to a specific loci

If we want all reads mapping within a specific genomic region, we can use `samtools view` and the `ref:start-end` syntax. You can use just the `ref` to extract an entire reference sequence such as a chromosome (example not shown here). This requires a BAM index file.

```{bash engine.opts='-l'}
samtools view eg/ERR188273_chrX.bam chrX:20000-30000
```

Note that this takes into account the mapping of the entire read and not just the starting position. For example, if you specified chrX:20000-30000, a read that is 75 bp long that maps to position 19999 will also be returned. You can save the output as another BAM file if you wish.

```{bash engine.opts='-l'}
samtools view -b eg/ERR188273_chrX.bam chrX:20000-30000 > eg/ERR188273_chrX_20000_30000.bam
```

You can also use a BED file, with several entries, to extract reads of interest.

```{bash engine.opts='-l'}
cat eg/my.bed 

samtools view -L eg/my.bed eg/ERR188273_chrX.bam
```

# Extracting only the first read from paired end BAM files

Sometimes you only want the first pair of a mate. 0x0040 is hexadecimal for 64 (i.e. 16 * 4), which is binary for 1000000, corresponding to the read in the first read pair.

```{bash engine.opts='-l'}
samtools view -b -f 0x0040 eg/ERR188273_chrX.bam > eg/first.bam
```

# Stats

For simple statistics use `samtools flagstat`.

```{bash engine.opts='-l'}
samtools flagstat eg/ERR188273_chrX.bam
```

For additional stats, use `samtools stats`.

```{bash engine.opts='-l'}
samtools stats eg/ERR188273_chrX.bam | grep ^SN
```

# Interpreting the BAM flags

The second column in a SAM/BAM file is the flag column. They may seem confusing at first but the encoding allows details about a read to be stored by just using a few digits. The trick is to convert the numerical digit into binary, and then use the table to interpret the binary numbers, where 1 = true and 0 = false. I wrote a blog post on BAM flags: <http://davetang.org/muse/2014/03/06/understanding-bam-flags/>, which also includes a Perl script for interpreting BAM flags. There is also the `flags` subcommand.

```{bash engine.opts='-l'}
samtools flags
```

# samtools calmd/fillmd

The calmd or fillmd tool is useful for visualising mismatches and insertions in an alignment of a read to a reference genome. The `-e` argument changes identical bases between the read and reference into `=`.

```{bash engine.opts='-l'}
samtools view -b eg/ERR188273_chrX.bam | samtools fillmd -e - /Users/dtang/github/rnaseq/raw/chrX_data/genome/chrX.fa > eg/ERR188273_chrX_fillmd.bam

head eg/ERR188273_chrX_fillmd.bam
```

# Creating fastq files from a BAM file

```{bash engine.opts='-l'}
samtools fastq -1 eg/ERR188273_chrX_1.fq -2 eg/ERR188273_chrX_2.fq eg/ERR188273_chrX.bam
head eg/ERR188273_chrX_1.fq
```

# Random subsampling of BAM file

The SAMtools view `-s` parameter allows you to randomly sample lines of a BAM file. Using `0.5` will subsample half of all mapped reads.

```{bash engine.opts='-l'}
samtools view -s 0.5 -b eg/ERR188273_chrX.bam > eg/ERR188273_chrX_rand.bam
```

# Count number of reads

Use `samtools idxstats` to print stats on a BAM file; this requires an index file which is created by running `samtools index`.

```{bash engine.opts='-l'}
# output of idxstats is:
# ref name, sequence length of ref, no. mapped reads, and no. unmapped reads
samtools idxstats eg/ERR188273_chrX.bam
```

We can use this with `awk` to sum up the columns.

```{bash engine.opts='-l'}
# number of reads = mapped + unmapped
samtools idxstats eg/ERR188273_chrX.bam | awk '{s+=$3+$4} END {print s}'

# number of mapped reads = 3rd column
samtools idxstats eg/ERR188273_chrX.bam  | awk '{s+=$3} END {print s}'
```

# Obtaining genomic sequence

Use `faidx` to fetch genomic sequence; coordinates are 1-based.

```{bash engine.opts='-l'}
# index fasta file
samtools faidx /Users/dtang/github/rnaseq/raw/chrX_data/genome/chrX.fa

# obtain sequence
samtools faidx /Users/dtang/github/rnaseq/raw/chrX_data/genome/chrX.fa chrX:300000-300100
```

# Comparing BAM files

Install [deepTools](https://deeptools.readthedocs.io/en/develop/) and use [bamCompare](https://deeptools.readthedocs.io/en/develop/content/tools/bamCompare.html). The bigWig output file shows the ratio of reads between `b1` and `b2` in 50 bp (default) windows.

# Converting reference names

One of the most annoying bioinformatics problems is the use of different chromosome names, e.g. chr1 vs 1, in different references even when the sequences are identical. The GRCh38 reference downloaded from Ensembl has chromosome names without the `chr`:

    >1 dna:chromosome chromosome:GRCh38:1:1:248956422:1 REF

Whereas the reference names from UCSC has the `chr`:

    >chr1  AC:CM000663.2  gi:568336023  LN:248956422  rl:Chromosome  M5:6aef897c3d6ff0c78aff06ac189178dd  AS:GRCh38

Luckily you can change the reference names using `samtools reheader` but just make sure your reference sequences are actually identical.

```{bash engine.opts='-l'}
samtools view eg/ERR188273_chrX.bam | head -2

# view header
samtools view -H eg/ERR188273_chrX.bam

# substitute header with new name
samtools view -H eg/ERR188273_chrX.bam | sed 's/SN:chrX/SN:X/' > eg/my_header

# save bam file with new ref
samtools reheader eg/my_header eg/ERR188273_chrX.bam > eg/ERR188273_X.bam

samtools view eg/ERR188273_X.bam | head -2
```

# Coverage

We can use `samtools depth` to tally the number of reads covering a region; the three columns are the reference, position, and read coverage. In the example below, there are two reads covering positions 200 - 205. The `samtools mpileup` command can provide more information, including:

1. Sequence name
2. 1-based coordinate
3. Reference base
4. Number of reads covering this position
5. Read bases
6. Base qualities
7. Alignment mapping qualities

See https://davetang.org/muse/2015/08/26/samtools-mpileup/ for more information.

```{bash engine.opts='-l'}
samtools depth eg/ERR188273_chrX.bam | head
```

# Stargazers over time

[![Stargazers over time](https://starchart.cc/davetang/learning_bam_file.svg)](https://starchart.cc/davetang/learning_bam_file)
